enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -3838,6 +3841,306 @@ class AdjointGenerator
       return;
     }
 
+    if (funcName == "cblas_ddot" && called->isDeclaration()) {
+      Type *castvals[2] = {call.getArgOperand(1)->getType(),
+                           call.getArgOperand(3)->getType()};
+      auto *cachetype = StructType::get(call.getContext(), ArrayRef(castvals));
+      Value *undefinit = UndefValue::get(cachetype);
+      Value *cacheval;
+      auto in_arg = call.getCalledFunction()->arg_begin();
+      in_arg++;
+      Argument *xfuncarg = in_arg;
+      in_arg++;
+      in_arg++;
+      Argument *yfuncarg = in_arg;
+      bool xoverwrite = uncacheable_args.find(xfuncarg)->second;
+      bool yoverwrite = uncacheable_args.find(yfuncarg)->second;
+      if ((Mode == DerivativeMode::ReverseModeCombined ||
+           Mode == DerivativeMode::ReverseModePrimal) &&
+          ((!gutils->isConstantValue(call.getArgOperand(3)) && xoverwrite) ||
+           (!gutils->isConstantValue(call.getArgOperand(1)) && yoverwrite))) {
+        auto size =
+            ConstantExpr::getSizeOf(Type::getDoubleTy(call.getContext()));
+        BuilderZ.SetInsertPoint(gutils->getNewFromOriginal(&call));
+        auto *newcall = dyn_cast<CallInst>(gutils->getNewFromOriginal(&call));
+        Value *arg1, *arg2;
+        if (!gutils->isConstantValue(call.getArgOperand(3)) && xoverwrite) {
+          auto malins1 = CallInst::CreateMalloc(
+              gutils->getNewFromOriginal(&call), size->getType(),
+              Type::getDoubleTy(call.getContext()), size, call.getArgOperand(0),
+              nullptr, "");
+          arg1 =
+              BuilderZ.CreateBitCast(malins1, call.getArgOperand(1)->getType());
+          auto memcpy_size1 =
+              (int)(call.getFunction()
+                        ->getParent()
+                        ->getDataLayout()
+                        .getTypeAllocSize(
+                            newcall->getArgOperand(1)->getType())) *
+              dyn_cast<ConstantInt>(call.getArgOperand(0))->getSExtValue();
+#if LLVM_VERSION_MAJOR >= 10
+          BuilderZ.CreateMemCpy(arg1, MaybeAlign(), newcall->getArgOperand(1),
```

Comment:
I'm a bit worried about the correctness of the memcpy here. The spec specifies a stride (which so far has been tested wiht only 1). But suppose an stride of 4 were used, we won't copy all the data needed. Arguably in that case we should still allocate the same amount but do something like cache[i] = input[4*i], then in the reverse should use a size of 1 for the cached version (since we're storing at stride of 1).

Reply:
Just to clarify,

For a strided implementation where we do something like - 
```
g(double *m, double* m1, double *n, double *n1) { // Where m1, n1 are shadow args
  cblas_ddot(3, m, 2, n, 3); // 2 is the stride for m. 3 is the stride for n.
  ....
}
  
int caller() {
  double m[6] = {1, 2, 3, 4, 5, 6};
  double m1[6] = {0, 0, 0, 0, 0, 0};
  double n[9] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
  double n1[9] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
  g(m, m1, n, n1);
```
If cblas_ddot's source code is available, we store the derivatives wrt m in m1[0], m1[2], m1[4]. And the derivatives wrt n in n1[0], n1[3], n1[6].

If I understand correctly, in the strided memcpy implementation that you suggest, we would be storing the derivatives wrt m in m1[0], m1[1], m1[2]. And the derivatives wrt n in n1[0], n1[1], n1[2]. Is that okay?

Reply:
So we still want the same end derivative -- but we don't need to cache the values in the same way. So even though here m has a stride of 2, we can condense the cache to have a stride of 1 when we copy into it. This should still produce the same result, so long as we now a stride of 1 when indexing into the cached m for the reverse.